File: Template/DataStructure/SparseTable.cpp
template<typename T>
struct SparseTable{
    vector<vector<T>> f_max;
    vector<vector<T>> f_min;
    vector<vector<T>> f_gcd;

    SparseTable(){}
    SparseTable(const vector<T> &v){
        init(v);
    }

    void init(const vector<T> &v){
        int n=v.size(),m=log2(n);
        f_max.assign(n,vector<T>(m+1));
        f_min.assign(n,vector<T>(m+1));
        f_gcd.assign(n,vector<T>(m+1));
        for(int j=0;j<=m;j++){
            for(int i=0;i+(1<<j)-1<n;i++){
                f_max[i][j]=(j!=0?max(f_max[i][j-1],f_max[i+(1<<(j-1))][j-1]):v[i]);
                f_min[i][j]=(j!=0?min(f_min[i][j-1],f_min[i+(1<<(j-1))][j-1]):v[i]);
                f_gcd[i][j]=(j!=0?gcd(f_gcd[i][j-1],f_gcd[i+(1<<(j-1))][j-1]):v[i]);
            }
        }
    }

    T query_max(int l,int r){
        int k=log2(r-l+1);
        return max(f_max[l][k],f_max[r-(1<<k)+1][k]);
    }

    T query_min(int l,int r){
        int k=log2(r-l+1);
        return min(f_min[l][k],f_min[r-(1<<k)+1][k]);
    }

    T query_gcd(int l,int r){
        int k=log2(r-l+1);
        return gcd(f_gcd[l][k],f_gcd[r-(1<<k)+1][k]);
    }
};

File: Template/DataStructure/SparseTable.py
class SparseTable:
    def __init__(self,arr):
        n=len(arr)
        self.LOG2=[-1]*(n+1)
        for i in range(1,n+1):
            self.LOG2[i]=self.LOG2[i>>1]+1
        N=int(log2(n))+1

        self.f_max=[[0]*N for i in range(n)]
        for j in range(N):
            i=0
            while i+(1<<j)-1<n:
                if not j:
                    self.f_max[i][j]=arr[i]
                else:
                    if self.f_max[i][j-1]>=self.f_max[i+(1<<j-1)][j-1]:
                        self.f_max[i][j]=self.f_max[i][j-1]
                    else:
                        self.f_max[i][j]=self.f_max[i+(1<<j-1)][j-1]
                i+=1

        # self.f_min=[[0]*N for i in range(n)]
        # for j in range(N):
        #     i=0
        #     while i+(1<<j)-1<n:
        #         if not j:
        #             self.f_min[i][j]=arr[i]
        #         else:
        #             if self.f_min[i][j-1]<=self.f_min[i+(1<<j-1)][j-1]:
        #                 self.f_min[i][j]=self.f_min[i][j-1]
        #             else:
        #                 self.f_min[i][j]=self.f_min[i+(1<<j-1)][j-1]
        #         i+=1

        # self.f_gcd=[[0]*N for i in range(n)]
        # for j in range(N):
        #     i=0
        #     while i+(1<<j)-1<n:
        #         if not j:
        #             self.f_gcd[i][j]=arr[i]
        #         else:
        #             self.f_gcd[i][j]=gcd(self.f_gcd[i][j-1],self.f_gcd[i+(1<<j-1)][j-1])
        #         i+=1

    def query_max(self,l,r):
        k=self.LOG2[r-l+1]
        if self.f_max[l][k]>=self.f_max[r-(1<<k)+1][k]:
            x=self.f_max[l][k]
        else:
            x=self.f_max[r-(1<<k)+1][k]
        return x

    # def query_min(self,l,r):
    #     k=self.LOG2[r-l+1]
    #     if self.f_min[l][k]<=self.f_min[r-(1<<k)+1][k]:
    #         x=self.f_min[l][k]
    #     else:
    #         x=self.f_min[r-(1<<k)+1][k]
    #     return x

    # def query_gcd(self,l,r):
    #     k=self.LOG2[r-l+1]
    #     x=gcd(self.f_gcd[l][k],self.f_gcd[r-(1<<k)+1][k])
    #     return x